cmake: Add `Coverage` script

hebasto opened the pull request:
Using separated scripts provides additional flexibility as there is no need to specify coverage report specific options during the build configuration stage.

Examples of usage on Ubuntu 24.04 with [LCOV v2.0](https://packages.ubuntu.com/noble/lcov):

- total coverage:
```
$ cmake -B build -DCMAKE_BUILD_TYPE=Coverage -DCMAKE_CXX_FLAGS="-fprofile-update=atomic"
$ cmake --build build
$ cmake -DLCOV_OPTS="--ignore-errors mismatch" -P build/Coverage.cmake
# OR
$ cmake \
    -DJOBS=$(nproc) \
    -DEXTENDED_FUNCTIONAL_TESTS=ON \
    -DLCOV_OPTS="--rc branch_coverage=1 --ignore-errors mismatch" \
    -P build/Coverage.cmake
```

- fuzz coverage:
```
$ cmake -B build -DCMAKE_BUILD_TYPE=Coverage -DCMAKE_CXX_FLAGS="-fprofile-update=atomic" -DBUILD_FUZZ_BINARY=ON
$ cmake --build build
$ cmake \
    -DFUZZ_SEED_CORPUS_DIR=<path/to/seed/corpus> \
    -DLCOV_OPTS="--rc branch_coverage=1 --ignore-errors mismatch" \
    -P build/CoverageFuzz.cmake
```

This PR supports both LCOV v1 and LCOV v2. Hence, it is possible to test it on Ubuntu 24.04.

hebasto commented:
Included changes from https://github.com/bitcoin/bitcoin/pull/30063.

hebasto commented:
@maflcko Want to tested this PR in your https://github.com/maflcko/b-c-cov?

hebasto commented on cmake/script/CoverageInclude.cmake.in:
```diff
@@ -0,0 +1,60 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
+if("@CMAKE_CXX_COMPILER_ID@" STREQUAL "Clang")
+  find_program(LLVM_COV_EXECUTABLE llvm-cov REQUIRED)
+  set(COV_TOOL "${LLVM_COV_EXECUTABLE} gcov")
```
Right. But this PR aims to mirror the master branch behaviour. Further improvements can be done later, no?

TheCharlatan commented on doc/developer-notes.md:
```diff
@@ -481,22 +481,27 @@ $ ./test/functional/test_runner.py --valgrind
-./configure --enable-lcov
-make
-make cov
+cmake -B build -DCMAKE_BUILD_TYPE=Coverage
```
> For example one may want to gather coverage on debug builds or release builds?

Is this something that is commonly done? What would the purpose of this be?

> I know this would be different from how autotools works, but I would propose to have a shell script

Not sure about this. I prefer to work with the build system directly without having to understand a wrapper script first if I want to tweak something.

hebasto commented:
1. Switched from the "Coverage" build type to the "coverage" preset. It improves flexibility and hopefully addresses @vasild's https://github.com/hebasto/bitcoin/pull/191#discussion_r1598375407.
2. Now supports both LCOV v1 and LCOV v2. Hence, it is possible to test on Ubuntu 24.04.

hebasto commented:
Friendly ping code coverage connoisseurs @maflcko @dergoegge :)

Your opinion will be much appreciated by all people in the CMake working group.

fanquake commented on cmake/script/CoverageInclude.cmake.in:
```diff
@@ -0,0 +1,60 @@
+  @ONLY
+)
+
+set(LCOV_OPTS --ignore mismatch)
```
I don't think this should be hardcoded here (it's not needed with lcov 1.x, and isn't the only ignore that might be required with 2.x)?

hebasto replied:
Thanks! [Deleted](https://github.com/hebasto/bitcoin/pull/191#issuecomment-2139141905).

fanquake commented on cmake/script/CoverageInclude.cmake.in:
```diff
@@ -0,0 +1,60 @@
+)
+
+set(LCOV_OPTS --ignore mismatch)
+foreach(setting IN LISTS LCOV_RC)
```
Is there are reason we've created a new external variable (`LCOV_RC`), rather than just using `LCOV_OPTS`, which is already available, and used?

hebasto replied:
Thanks! [Reworked](https://github.com/hebasto/bitcoin/pull/191#issuecomment-2139141905).

hebasto commented:
Addressed the recent @fanquake's comments.

dergoegge commented on doc/developer-notes.md:
```diff
@@ -481,22 +481,35 @@ $ ./test/functional/test_runner.py --valgrind
-./configure --enable-lcov
-make
-make cov
+cmake -B build --preset coverage
```
This is different from the examples in the PR description and also skips building the fuzz binary, is that expected?

hebasto replied:
The PR description has been updated.

hebasto replied:
UPD. Added `--ignore-errors mismatch`

hebasto commented:
Re-tested, fixed bugs.

Reverted back to use of the "Coverage" build type because this guarantees that the `Coverage.cmake` script finds unit tests when using multi-config generators.

The PR description has been updated.

fanquake commented on CMakeLists.txt:
```diff
@@ -484,6 +484,22 @@ else()
   unset(c_flags_debug_overridden)
 endif()
 
+set(CMAKE_CXX_FLAGS_COVERAGE "-Og --coverage -fprofile-update=prefer-atomic")
```
> fprofile-update=prefer-atomic

Not sure this should be in here? Don't think it's needed with Clang, and only sometimes with GCC?

hebasto replied:
Would it hurt if applied in such cases? Or do you want me to remove it?

hebasto replied:
> Not sure this should be in here?

Deleted.

The PR description has been updated.

dergoegge commented:
```
Writing data to fuzz_filtered.info
Summary coverage rate:
  lines......: 20.2% (19593 of 97038 lines)
  functions..: 20.4% (18829 of 92323 functions)
  branches...: no data found
Reading data file fuzz_coverage.info
Found 1190 entries.
Found common filename prefix "/home/dergoegge/workspace/bitcoin"
Writing .css and .png files.
Generating output.
genhtml: ERROR: cannot read /home/dergoegge/workspace/bitcoin/build/src/include/boost/visit_each.hpp
Processing file build/src/include/boost/visit_each.hpp
```

My lcov version is quite old (1.0), maybe that's why this fails. I've had to fiddle with the versions in the past and I'm sure this would work with the right versions etc. (everything seemed to work here besides the actual html generation).

I've switched to using llvm's native coverage reporting (i.e. compiling with `-fprofile-instr-generate -fcoverage-mapping`) which has been a breeze so far. I switched because I kept running into weird lcov problems like this and the gcov reports are also not the best (e.g. inline branch coverage is reported in a really weird format and I can't find any docs).

hebasto commented:
Rebased.

hebasto commented:
> In the first `OR` in the pull request description in the second branch, you forgot to specify `DCMAKE_CXX_FLAGS` and the build type?

This `OR` is related to the script invocation:
```
$ cmake -DLCOV_OPTS="--ignore-errors mismatch" -P build/Coverage.cmake
# OR
$ cmake \
    -DJOBS=$(nproc) \
    -DEXTENDED_FUNCTIONAL_TESTS=ON \
    -DLCOV_OPTS="--rc branch_coverage=1 --ignore-errors mismatch" \
    -P build/Coverage.cmake
```

vasild commented:
Further improvements, that would deviate from autotools:

* Reconsider `-Og` because it has different meanings for [GCC](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-Og) and [Clang](https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-O0). Optimizations could result in bogus line numbers in the output, maybe just use `-O0 -g`.
* Rename `cmake/script/Coverage.cmake` to `GccCoverage.cmake` because Clang uses a different workflow: https://clang.llvm.org/docs/SourceBasedCodeCoverage.html. Introduce `ClangCoverage.cmake`.
* Drop the build target `Coverage`. Passing extra compiler flags is more suitable for a CMake option, rather than a build target. IMO that is very similar to `-DWERROR=ON` for example. Introduce `-DCOVERAGE=OFF|GCC|CLANG`.
